net: buf: linearize: Never return -ENOMEM, just do what user asked to

pfalcon opened the pull request:
Don't try to find "errors" in the values of dst_len and len params
passed to net_buf_linearize(). Instead, do what entails with the
common sense from the values passed in, specifically:

1. Never read more than dst_len (or it would lead to buffer
overflow).
2. It's absolutely ok to read than specified by "len" param, that's
why this function returns number of bytes read in the first place.

The motivation for this change is that it's not useful with its
current behavior. For example, a number of Ethernet drivers linearize
a packet to send, but each does it with its own duplicated adhoc
routine, because net_buf_linearize() would just return error for the
natural use of:

net_buf_linearize(buf, sizeof(buf), pkt->frags, 0, sizeof(buf));

Signed-off-by: Paul Sokolovsky <paul.sokolovsky@linaro.org>

pfalcon committed:
net: buf: linearize: Never return -ENOMEM, just do what user asked to

Don't try to find "errors" in the values of dst_len and len params
passed to net_buf_linearize(). Instead, do what entails with the
common sense from the values passed in, specifically:

1. Never read more than dst_len (or it would lead to buffer
overflow).
2. It's absolutely ok to read than specified by "len" param, that's
why this function returns number of bytes read in the first place.

The motivation for this change is that it's not useful with its
current behavior. For example, a number of Ethernet drivers linearize
a packet to send, but each does it with its own duplicated adhoc
routine, because net_buf_linearize() would just return error for the
natural use of:

net_buf_linearize(buf, sizeof(buf), pkt->frags, 0, sizeof(buf));

Signed-off-by: Paul Sokolovsky <paul.sokolovsky@linaro.org>

pfalcon commented:
Currently posted to ask CI's opinion.

Vudentz commented:
I don't thing using #if 0 is desirable in any circustance, or perhaps you are just using it in order to highlight the code that should be removed? Other than that this looks good to me, in fact even if there were legitimate reason to check for no memory that should probably be converted to assert since that is an obvious bug in the code.

pfalcon commented:
@Vudentz: Ack, posted just to see if there would be any CI failures, and get any possible quick comments from reviewers. Seeing no failures/immediate objections, will definitely update to properly remove the affected code.

Thanks for looking! Happy New Year!

tbursztyka commented on subsys/net/buf.c:
```diff
@@ -681,9 +681,12 @@ int net_buf_linearize(void *dst, size_t dst_len, struct net_buf *src,
 	size_t copied;
 
+	len = min(len, dst_len);
+#if 0
```
why not just removing the old code instead?

tbursztyka reviewed (dismissed):
net_buf.h documentation would need to be updated as well.

pfalcon committed:
net: pkt: net_frag_linearize: Make consistent with net_buf_linearize

net_frag_linearize() is just a wrapper for net_buf_linearize(). As
the latter was refactored to never return error, and instead just
return actual copied length, update the former and its usages too.

Signed-off-by: Paul Sokolovsky <paul.sokolovsky@linaro.org>

pfalcon commented:
@Vudentz, @tbursztyka: Finished refactoring, please re-review. Suggestions for squashing commits together are accepted.

pfalcon committed:
tests: net_pkt: More net_frag_linearize tests

Add tests for "can return less data than requested" behavior.

Signed-off-by: Paul Sokolovsky <paul.sokolovsky@linaro.org>

pfalcon commented:
Added tests for the updated function behavior.

pfalcon commented:
@tbursztyka : Ping

tbursztyka commented on include/net/net_pkt.h:
```diff
@@ -1122,18 +1122,16 @@ int net_frag_linear_copy(struct net_buf *dst, struct net_buf *src,
- * @return -ENOMEM on error
+ * @return number of bytes actually copied
  */
 int net_frag_linearize(u8_t *dst, size_t dst_len,
```
Shouldn't it return size_t as well then?

jhedberg replied:
ssize_t then, I suppose, since it needs to return negative values as well.

jhedberg replied:
Nevermind, I see the negative errors are removed by this PR

pfalcon commented:
Thanks for the reviews. I squashed commits down to 3 after all, so now there's a commit to update net_buf_linearize(), a commit to update net_frag_linearize(), and a commit to add more tests for net_buf_linearize().

pfalcon commented on subsys/net/buf.c:
```diff
@@ -674,16 +674,14 @@ struct net_buf *net_buf_frag_del(struct net_buf *parent, struct net_buf *frag)
+	len = min(len, dst_len);
 
 	frag = src;
 
```
> Unrelated to this patch ...

Ah, don't ask, I see at least one another "unrelated" thing to this patch. Step by step...

nashif commented on include/net/buf.h:
```diff
@@ -1300,21 +1300,22 @@ struct net_buf *net_buf_frag_del(struct net_buf *parent, struct net_buf *frag);
 
 /**
- * @brief Copy len bytes from src starting from offset to dst buffer
+ * @brief Copy len bytes from net_buf chain starting at offset to linear buffer
```
remove len here -> "Copy bytes ..."
